src/test/java/uk/nhs/hee/tis/trainee/credentials/service/CachingDelegateIntegrationTest.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright 2023 Crown Copyright (Health Education England)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
+ * associated documentation files (the "Software"), to deal in the Software without restriction,
+ * including without limitation the rights to use, copy, modify, merge, publish, distribute,
+ * sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
+ * NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package uk.nhs.hee.tis.trainee.credentials.service;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+import java.time.LocalDate;
+import java.util.Optional;
+import java.util.UUID;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.annotation.DirtiesContext;
+import org.springframework.test.annotation.DirtiesContext.ClassMode;
+import org.springframework.test.context.ActiveProfiles;
+import org.testcontainers.junit.jupiter.Testcontainers;
+import uk.nhs.hee.tis.trainee.credentials.dto.IdentityDataDto;
+
+@SpringBootTest
+@ActiveProfiles("redis")
+@DirtiesContext(classMode = ClassMode.AFTER_EACH_TEST_METHOD)
```

Comment:
I can see the value in starting each test with a clean slate, but I am finding this means I end up with 19 redis containers running. If I remove this line, I end up with 4, and the tests seem to run a fair bit quicker as well. They still all pass, which I guess is due to an absence of collisions from the UUID.randomUUID() key used in each test. Given that these tests seem to replicate the ones in the CachingDelegateTest, but specifically against a redis cache, I wonder if there is a way to simplify / speed-up things by either not entirely duplicating all the tests here, or working on the assumption that the keys won't collide and hence not worrying about cleaning the context after each test?

Reply:
`CacheDelegateTest` is more of a limited fallback for when Docker isn't available, it can only check what the hardcoded returns for the methods and not what caching annotations are set up and do.

I'll have a look at the context dirtying, you're right that we shouldn't get any collisions anyway - so the less redis instances the better.